### PR TITLE
Fix png color metadata loss

### LINKED ISSUE #200: ICC color profiles are stripped from profile photo uploads (without sRGB conversion)
When uploading a profile photo that uses ICC colorspace "Display P3", Dinopark simply strips the ICC profile without converting the image to sRGB. This results in damaged profile photos for anyone uploading photos from a wide color pipeline - such as modern mobile phones, DSLR cameras, or any Apple devices.

Please either retain the ICC profile included with the user's uploaded image, or convert the uploaded image to sRGB ('relative colormetric') prior to stripping the ICC profile from the image.

An example of this issue can be viewed on my profile page, where I uploaded a P3-tagged image and the image currently displayed demonstrates RGB values from the P3 colorspace but is displayed incorrectly in Firefox due to its assumption that untagged images are sRGB. I am available on Slack to help support analysis and repair of this issue.

### PR DESCRIPTION
**Pull request by jens1o**
Fixes https://github.com/mozilla-iam/dino-park-issues/issues/200

- [x] Add tests

### CODE REVIEW

File: src/send/resize.rs [hunk 1]
```diff
@@ -14,25 +16,92 @@ pub struct Avatars {
...
+
+                metadata_to_add
+                    .write_u32::<byteorder::BE>(data_chunk.crc())
+                    .unwrap();
+
+                added_chunks.push(chunk_to_copy);
+            }
+        }
+
+        info!(
```
Comment (fiji-flo) [inline src/send/resize.rs]: lowercase and move to `debug!` maybe add a generic `info!("png metadata copied")` or something like that.

File: src/send/resize.rs [hunk 2]
```diff
@@ -14,25 +16,92 @@ pub struct Avatars {
...
 }
 
-fn downsize(size: u32, img: &DynamicImage) -> Result<Vec<u8>, Error> {
+fn downsize(size: u32, img: &DynamicImage, metadata_to_add: &[u8]) -> Result<Vec<u8>, Error> {
     let down_sized = img.resize_to_fill(size, size, FilterType::Lanczos3);
     let mut buf: Vec<u8> = Vec::new();
     down_sized.write_to(&mut buf, image::ImageOutputFormat::Png)?;
+
+    if !metadata_to_add.is_empty() {
+        const IMAGE_START_OFFSET_TO_ADD_DATA: usize = 33;
```
Comment (fiji-flo) [inline src/send/resize.rs:91]: Maybe add a comment why 33 bytes offset.

File: src/send/resize.rs [hunk 3]
```diff
@@ -14,25 +16,92 @@ pub struct Avatars {
...
+
+        // temporarily cut off the image bytes after the (fixed) header
+        // as most of the metadata we need to add must appear before any other tag
+        let image_data = &buf[IMAGE_START_OFFSET_TO_ADD_DATA..]
+            .iter()
+            .copied()
+            .collect::<Vec<u8>>();
+
+        buf.truncate(IMAGE_START_OFFSET_TO_ADD_DATA);
```
Comment (fiji-flo) [inline src/send/resize.rs]: [split_off](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.split_off) does this I think.

File: src/send/resize.rs [hunk 4]
```diff
@@ -14,25 +16,92 @@ pub struct Avatars {
...
+            x528: downsize(528, &img, &metadata_to_add)?,
+            x264: downsize(264, &img, &metadata_to_add)?,
+            x100: downsize(100, &img, &metadata_to_add)?,
+            x40: downsize(40, &img, &metadata_to_add)?,
         })
     }
+
+    /// Returns png chunks that needs to be copied to keep color information
+    /// related things intact (i.e. copy chunks that the image crate does not pick up)
+    fn maybe_extract_png_color_metadata(buf: &[u8]) -> Result<Vec<u8>, Error> {
```
Comment (fiji-flo) [inline src/send/resize.rs:48]: Could also look like this:
```rust
    fn maybe_extract_png_color_metadata(buf: &[u8]) -> Result<Vec<u8>, Error> {
        let mut png_decoder = lodepng::Decoder::new();
        // required so the chunks are remembered (by default, it only remembers the mandatory chunks)
        png_decoder.remember_unknown_chunks(true);
        png_decoder.decode(&buf)?;

        let metadata_to_add = ["cHRM", "gAMA", "sRGB", "iCCP", "eXIf"]
            .iter()
            .filter_map(|chunk_to_copy| png_decoder.info_png().get(chunk_to_copy))
            .map(Avatars::metadata)
            .flatten()
            .collect();

        Ok(metadata_to_add)
    }

    fn metadata(data_chunk: lodepng::ChunkRef) -> Vec<u8> {
        // chunk_length = padding (2 bytes) chunk_len (4 bytes) + chunk type (4 bytes) + data_len + crc (8 bytes) = data_len + 18
        let mut metadata_to_add = Vec::with_capacity(data_chunk.len() + 18);

        // write two bytes of padding
        metadata_to_add.extend(&[0x00, 0x00]);
        metadata_to_add
            .write_u16::<byteorder::BE>(data_chunk.len() as u16)
            .unwrap();
        metadata_to_add.extend(&data_chunk.name());
        metadata_to_add.extend(data_chunk.data());

        metadata_to_add
            .write_u32::<byteorder::BE>(data_chunk.crc())
            .unwrap();

        metadata_to_add
    }
```

And moving the `info!` to the place we add the metadata. Optionally `debug!` log which chunks we copy in then `map`

**Review by fiji-flo — CHANGES_REQUESTED** [review]
Only two things are the verbose info logging and a comment for the magic 33. Everything else is up to you.

Great work!

**Comment by jens1o** [comment]
@fiji-flo I just forgot to remind you about the S3 dump of the raw images regarding the post-processing and generation of new downsized images. Could you prepare that for me, please?

**Review by fiji-flo — APPROVED** [review]
Nice work.

**Pull request merged**